app/services/auth.server.ts

```diff
@@ -0,0 +1,204 @@
+  const decodedToken = decode(accessToken, { complete: true, json: true })
+
+  try {
+    const rfd = client(accessToken)
+    const apiUser = handleApiResponse(await rfd.methods.getSelf({}))
+    const groups = handleApiResponse(await await rfd.methods.getGroups({}))
+
+    const emails = [...new Set(apiUser.providers.flatMap((provider) => provider.emails))]
+    const email = emails[0]
+    const displayNames = [
+      ...new Set(apiUser.providers.flatMap((provider) => provider.displayNames)),
+    ]
+    const displayName = displayNames[0]
+
+    if (typeof decodedToken?.payload === 'string') {
+      throw new Error('Invalid token payload')
+    }
+
+    const expiresAt = new Date((decodedToken?.payload.exp ?? 0) * 1000)
+
+    let user: User = {
+      id: apiUser.info.id,
+      email,
+      displayName,
+      token: accessToken,
+      permissions: apiUser.info.permissions,
+      groups: apiUser.info.groups
+        .map((groupId) => {
+          return groups.find((group) => group.id === groupId)?.name
+        })
+        .filter(isTruthy),
+      expiresAt,
+    }
+
+    return user
+  } catch (err) {
+    console.error('Failed to generate user', err)
+    throw err
+  }
+}
+
+const verify: RfdVerifyCallback<User> = async ({ tokens }) => {
+  const accessToken = tokens.accessToken()
+  return fetchUser(accessToken)
+}
+
+const googleOAuth = new RfdOAuthStrategy(
+  {
+    host: process.env.RFD_API || '',
+    clientId: process.env.RFD_API_CLIENT_ID || '',
+    clientSecret: process.env.RFD_API_CLIENT_SECRET || '',
+    redirectURI: process.env.RFD_API_GOOGLE_CALLBACK_URL || '',
+    remoteProvider: 'google',
+    scopes: ['group:info:r', 'rfd:content:r', 'rfd:discussion:r', 'search', 'user:info:r'],
+  },
+  verify,
+)
+auth.use(googleOAuth)
+
+const githubOAuth = new RfdOAuthStrategy(
+  {
+    host: process.env.RFD_API || '',
+    clientId: process.env.RFD_API_CLIENT_ID || '',
+    clientSecret: process.env.RFD_API_CLIENT_SECRET || '',
+    redirectURI: process.env.RFD_API_GITHUB_CALLBACK_URL || '',
+    remoteProvider: 'github',
+    scopes: ['group:info:r', 'rfd:content:r', 'rfd:discussion:r', 'search', 'user:info:r'],
+  },
+  verify,
+)
+auth.use(githubOAuth)
+
+const magicLink = new RfdMagicLinkStrategy(
+  {
+    storage: sessionStorage,
+    host: process.env.RFD_API || '',
+    clientSecret: process.env.RFD_API_MLINK_SECRET || '',
+    pendingPath: '/login?email=sent',
+    returnPath: '/auth/magic/callback',
+    channel: 'login',
+    linkExpirationTime: 60 * 10, // 10 minutes (in seconds)
+    scope: ['group:info:r', 'rfd:content:r', 'rfd:discussion:r', 'search', 'user:info:r'],
+  },
+  async ({ token }) => {
+    return fetchUser(token)
+  },
+)
+auth.use(magicLink)
+
+export async function getUserFromSession(request: Request): Promise<User | null> {
+  const session = await sessionStorage.getSession(request.headers.get('Cookie'))
+  const user = session.get('user')
+  return user
+}
```

Comment:
This is only used once outside this file, so I'd inline the contents into `authenticate` and just use that everywhere.